### ISSUE TITLE
Add new time operators

### ISSUE THREAD

**Issue by pauldix** [influxdata/influxdb#2071]
We should give users the ability to query times using operators other than `now()`. These would be some good additions:
- today() - sets time to now() rounded down to start of day
- yesterday() - sets time to now() - 1d rounded down to start of day
- monday() - the start of the previous monday(). Do this for every day of the week
- week() - the start of this week. Could take an argument for what day the week starts on
- month() - month boundaries

This won't go into 0.9.0, but can make it into a point release. Unless someone in the community wants to get crazy and implement it. For the adventurous you should look in these two places:
https://github.com/influxdb/influxdb/blob/master/influxql/engine.go#L583
https://github.com/influxdb/influxdb/blob/master/influxql/ast.go#L2192-L2200

**Comment by corylanou** [comment]
How do you want to handle timezones?

**Comment by corylanou** [comment]
Oh yea, and Daylight Savings Time :-).

**Comment by nemesisdesign** [comment]
could influxdb use the date set on the server? Or would you rather put it in the config file?

**Comment by XANi** [comment]
or just have separate operator for it for example `today(tz("PDT"))`  or more generic `today() + tz("PDT")`

That could also be handy for using "in reverse" for example if for some reason input data TS was in certain timezone, calculating GMT time would just require `field - tz("PDT")`

Or converting between them, say from PST to CET: `field - tz("PDT") + tz("CET")`

**Comment by pauldix** [comment]
I almost feel that the timezone could be a function argument on the end of the query. Since it can be applied to either the group by time or the relative time functions. And you'd always want it the same for both. So maybe:

```
select mean(value) from cpu
where time >= today()
group by time(10m)
tz("PST")
```

Where the `tz` function could take the string name of a time zone or it could take a positive or negative integer for offset from UTC.

**Comment by toddboom** [comment]
I agree with @pauldix on making it a function argument. I think doing time zone arithmetic is a little overkill, and can be solved by doing the explicit integer difference (i.e. +3/-3). Can we call it `time_zone` instead of `tz`?

This already came up in #388, but I'm thinking we should make a new issue to keep the discussion clean. Moving this to #2074.

**Comment by atombender** [comment]
I just encountered the problem of week binning: InfluxDB seems to assume to weeks start on Sundays, but the ISO standard and much of the world uses Monday (the US is an outlier). This needs to be configurable.

The fact that `time` is synctactically expressed as a partially-applied function makes this a bit awkward unless you introduce position-independent named arguments. How about taking a leaf out of SQL's book and introducing some readable keywords?

```sql
...
group by time
  time_zone 'CET'
  bin_by '1w'
  week_start 'monday'
```

**Comment by pqvst** [comment]
A good source for inspiration might be the moment.js library (I find it very intuitive and simple when it comes to parsing and manipulating time). http://momentjs.com/docs/#/manipulating/. 

The ``startOf`` and ``endOf`` operators on dates would be a very generic approach to calculating timespans like "this month", "last month", etc.

These could be implemented as alternatives to ``now()``:

* ``now()`` => current date/time
* ``day()`` => start of current day (2015-10-08 00:00:00)
* ``month()`` => start of current month (2015-10-01 00:00:00)
* ``year()`` => start of current year (2015-01-01 00:00:00)
* etc...

Now we can easily build more complex time spans like:

* today: ``time >= day() and time < day() + 1d``
* yesterday: ``time >= day() - 1d and time < day()``
* this month: ``time >= month() and time < month() + 1m``
* this day last month: ``time >= day() - 1m and time < day() - 1m + 1d``

In these cases, however, addition and subtraction of time/dates would have to be correct (due to leap-time, varying number of days in months, and all other messy date/time things). I'm not sure how that is implemented at the moment.

``month() + 1m`` is not equivalent to ``month() + 30d``
``day() + 1d`` is not equivalent to ``day() + 24h``
``...``

Furthermore, some of these time ranges could be implemented for us:

``last_month()`` => ``time >= ... and time < ...``

I might take a look at this if no work is being done on it. @pauldix, the first link you suggested is a 404. Has the filename changed?

**Comment by stephanstapel** [comment]
+1

**Comment by mvadu** [comment]
Was it partially implemented some time back? I am running against 0.9.5, and can use `time > today()` in my queries.

**Comment by Lahorde** [comment]
+1

Now, how can I do to get a count of values for today?

If I do :

    select count(value) from series where time >= now() - 1d
    -----------------------------------------
    time                            count
   2016-06-15T11:53:17.649981214Z  34

I won't get count from 00:00

If I do :

    select count(value) from series where time >= now() - 1d group by time(1d)
    -----------------------------------------
    time                    count
    2016-06-15T00:00:00Z    12
    2016-06-16T00:00:00Z    22

I will get two points as result
Correct value is 22 but how can I select only this value using influxdb?

**Comment by jwilder** [comment]
Related #1154 #3290

**Comment by kostasb** [comment]
Additional time operators should also take part in `GROUP BY time` clauses, such us `group by time(1M)` for monthly grouping. 
This will enable queries to group results using actual calendar broundaries.
E.g. `group by time(30d)` is not necessarily equivalent to 1 calendar month, depending on how long a calendar month is.

**Comment by nirpodo** [comment]
+1

**Comment by warezWally** [comment]
+1

Has any traction been made on this? I Would adore this feature.

**Comment by jsternberg** [comment]
Related to #6541 and I have a PR ready to add some of these time operators after #7762 is merged.

**Comment by ryanmills** [comment]
+1 for time(1M)

**Comment by krise3k** [comment]
@jsternberg #7762 has been merged, so I send kindly reminder about this feature :)

**Comment by willguxy** [comment]
+1 and would be great to have time of the day sort of operations. e.g.

- hour()
- minute()
- second()

For `2017-01-01 10:03:21` the above should return 10, 3, 21 respectively

**Comment by omki2005** [comment]
+1 Also would appreciate having time of the day operations. We want to be able to group data between say 9AM and 10AM and apply a different tag to "hour-of-day"

**Comment by Naesstrom** [comment]
+1 on this, would love to get yesterdays max value for example!

**Comment by aldas** [comment]
to be able to use relative time for at least 'start of today' (UTC is fine) in query would be really helpful. Atm you need to calculate this stuff in application code and use absolute time in query.

**Comment by sebastien85** [comment]
+1

**Comment by MitchVL** [comment]
+1

**Comment by liangshengzh** [comment]
+1

**Comment by superlevure** [comment]
Is there any news on this feature request ?

**Comment by alexanderturner** [comment]
+1

**Comment by noppelmax** [comment]
+1

**Comment by jpldev** [comment]
+1

**Comment by sibbl** [comment]
I'm not sure what I dislike more, these "+1" messages or the stale bot closing issues if you don't write "+1" 🙄

Please reopen this, I'd love to have time operators and keep this tracked here.

**Comment by jsternberg** [comment]
I apologize that the closing of this issue happened because of a bot, but the current plan is these will be implemented within Flux and there are no current plans to implement this in influxql.

**Comment by anderl80** [comment]
I'd be delighted to have more time operators, too!

**Comment by ThorbjornSomod** [comment]
Any progress on this? Will this be a feature of Flux? If yes, when will the feature be introduced? I see it is not a part of Flux as of yet.

**Comment by ccall48** [comment]
Would also love this to be added as a feature.

**Comment by jsternberg** [comment]
Going to be tracking this here: https://github.com/influxdata/flux/issues/1549
